core/profiles/src/main/java/software/amazon/awssdk/profiles/internal/ProfileFileRefresher.java

```diff
@@ -0,0 +1,267 @@
+                                                                                                 .build();
+    private final CachedSupplier<ProfileFileRefreshRecord> profileFileCache;
+    private volatile ProfileFileRefreshRecord currentRefreshRecord;
+    private final Supplier<ProfileFile> profileFile;
+    private final Path profileFilePath;
+    private final Function<RuntimeException, ProfileFile> exceptionHandler;
+    private final Consumer<ProfileFile> onProfileFileReload;
+    private final Clock clock;
+
+    private ProfileFileRefresher(Builder builder) {
+        this.exceptionHandler = builder.exceptionHandler;
+        this.clock = builder.clock;
+        this.profileFile = builder.profileFile;
+        this.profileFilePath = builder.profileFilePath;
+        this.onProfileFileReload = builder.onProfileFileReload;
+        this.profileFileCache = CachedSupplier.builder(this::refreshResult)
+                                              .clock(this.clock)
+                                              .build();
+        this.currentRefreshRecord = EMPTY_REFRESH_RECORD;
+    }
+
+    /**
+     * Builder method to construct instance of ProfileFileRefresher.
+     */
+    public static ProfileFileRefresher.Builder builder() {
+        return new ProfileFileRefresher.Builder();
+    }
+
+    /**
+     * Retrieves the cache value or refreshes it if stale.
+     */
+    public ProfileFile refreshIfStale() {
+        ProfileFileRefreshRecord cachedOrRefreshedRecord = profileFileCache.get();
+        ProfileFile cachedOrRefreshedProfileFile = cachedOrRefreshedRecord.profileFile;
+        if (isNewProfileFile(cachedOrRefreshedProfileFile)) {
+            currentRefreshRecord = cachedOrRefreshedRecord;
+        }
+
+        return cachedOrRefreshedProfileFile;
+    }
+
+    @Override
+    public void close() {
+        profileFileCache.close();
+    }
+
+    private RefreshResult<ProfileFileRefreshRecord> refreshResult() {
+        try {
+            return reloadAsRefreshResultIfStale();
+        } catch (RuntimeException exception) {
+            Instant now = Instant.now();
+            Instant staleTime = now;
+            ProfileFile exceptionProfileFile = exceptionHandler.apply(exception);
+            ProfileFileRefreshRecord refreshRecord = ProfileFileRefreshRecord.builder()
+                                                                             .profileFile(exceptionProfileFile)
+                                                                             .refreshTime(now)
+                                                                             .build();
+
+            return wrapIntoRefreshResult(refreshRecord, staleTime);
+        }
+    }
+
+    private RefreshResult<ProfileFileRefreshRecord> reloadAsRefreshResultIfStale() {
+        Instant now = clock.instant();
+        Instant staleTime = now;
```

Comment:
Why have two variables for the same value?

Reply:
Was being too verbose. Remove this and other instance.